File: py/horus/sdk/health.py
import dataclasses
from datetime import datetime
import enum
import typing
import horus.pb.project_manager.service_pb2 as pm_pb
import horus.pb.license_server.messages_pb2 as lc_pb
from horus.logs.format import (
    format_license_active_info,
    format_license_expired_error,
    format_license_not_found_error,
    format_license_server_connection_error,
    format_service_connection_timed_out,
)
from horus.sdk.sensor import SensorInfo, SensorStatus


@dataclasses.dataclass(frozen=True)
class LicenseInfo:
    number_of_lidars: int
    """Number of simultaneous lidars allowed by the license."""
    expiration_epoch: datetime
    """Timestamp of license expiration."""

    @staticmethod
    def _from_pb(pb: lc_pb.LicenseInfo) -> "LicenseInfo":
        return LicenseInfo(
            number_of_lidars=pb.lidar_count,
            expiration_epoch=datetime.fromtimestamp(pb.expiration_date.seconds),
        )


class LicensePrivilege(enum.Flag):
    """Privilege granted by a license."""

    LOWEST = 1
    SUBSCRIBE = 2
    LICENSE_ACTIONS = 4
    READ = 8
    WRITE = 16


@dataclasses.dataclass(frozen=True)
class LicenseStatus:
    """License status of horus instance."""

    level: int
    """Numeric representation of the current level. Ranges from 1 to 4."""
    reason: str
    """The reason why the license is at that level."""
    privileges: typing.FrozenSet[LicensePrivilege]
    """The set of privileges the current license grants you."""
    license_info: typing.Optional[LicenseInfo]
    """If a license was found, i.e if level is >= 3, contains the info of the license."""

    @staticmethod
    def _from_pb(pb: pm_pb.LicenseStatus) -> "LicenseStatus":
        level: int
        reason: str
        privileges: typing.FrozenSet[LicensePrivilege]
        license_info: typing.Optional[LicenseInfo] = None

        level_oneof = pb.license_level.WhichOneof("level")
        if level_oneof == "level_1":
            level = 1
            reason = format_license_server_connection_error(
                pb.license_level.level_1.license_server_unreachable
            )
        elif level_oneof == "level_2":
            level = 2
            reason = format_license_not_found_error(
                pb.license_level.level_2.license_not_found
            )
        elif level_oneof == "level_3":
            level = 3
            reason_oneof = pb.license_level.level_3.WhichOneof("reason")
            if reason_oneof == "license_expired":
                reason = format_license_expired_error(
                    pb.license_level.level_3.license_expired
                )
            elif reason_oneof == "license_server_disconnected":
                reason = format_license_server_connection_error(
                    pb.license_level.level_3.license_server_disconnected
                )
            else:
                raise Exception("Invalid LicenseLevel enum value")
            license_info = LicenseInfo._from_pb(pb.license_info)
        elif level_oneof == "level_4":
            level = 4
            reason = format_license_active_info(pb.license_level.level_4.license_active)
            license_info = LicenseInfo._from_pb(pb.license_info)
        else:
            raise Exception("Invalid LicenseLevel enum value")

        privileges = frozenset(
            LicensePrivilege(privilege)
            for privilege in pm_pb.LicensePrivilege.values()
            if privilege != 0 and (pb.license_level.privilege & privilege) == privilege
        )
        return LicenseStatus(
            level=level,
            reason=reason,
            privileges=privileges,
            license_info=license_info,
        )


@dataclasses.dataclass(frozen=True)
class SensorHealth:
    """A sensor health status. Maybe the node serving the sensor info was unreachable."""

    lidar_id: str
    """Unique identifier of the sensor."""
    status: typing.FrozenSet[SensorStatus]
    """Status of the sensor."""
    measured_frequency_hz: float
    """Frequency at which the sensor is measuring."""

    node_unreachable_error: typing.Optional[str]

    def is_unreachable(self) -> bool:
        """Returns whether the sensor could not be fetched from the preprocessing node."""
        return self.node_unreachable_error is not None

    @staticmethod
    def _from_pb(pb: pm_pb.GetHealthStatusResponse.SensorHealth) -> "SensorHealth":
        sensor_info = SensorInfo._from_pb(pb.info)

        node_unreachable_error: typing.Optional[str] = None
        if pb.WhichOneof("node_unreachable_error") == "timeout":
            node_unreachable_error = format_service_connection_timed_out(pb.timeout)

        return SensorHealth(
            lidar_id=sensor_info.lidar_id,
            status=sensor_info.status,
            measured_frequency_hz=sensor_info.measured_frequency_hz,
            node_unreachable_error=node_unreachable_error,
        )


@dataclasses.dataclass(frozen=True)
class NodeHealth:
    """Connectivity status of a specific node."""

    class Service(enum.Enum):
        """Service represented by a node."""

        DETECTION = 1
        LIDAR_RUNNER = 2
        NOTIFICATION = 3
        POINT_AGGREGATOR = 4
        PREPROCESSING = 5
        PROJECT_MANAGER = 6

    class Status(enum.Enum):
        """Connectivity status of a node."""

        UNREACHABLE = 1
        ALIVE = 2

    service: Service
    """Service represented by the node."""
    node_id: str
    """Unique node identifier."""
    node_status: Status
    """Connectivity status."""

    @staticmethod
    def _from_pb(
        pb: pm_pb.GetHealthStatusResponse.NodeHealth,
    ) -> "NodeHealth":
        return NodeHealth(
            node_id=pb.node_id,
            node_status=NodeHealth.Status(pb.node_status),
            service=NodeHealth.Service(pb.service),
        )


@dataclasses.dataclass(frozen=True)
class HealthStatus:
    """Response of `ProjectManager.GetHealthStatus`."""

    license_status: LicenseStatus
    """Status of the license."""
    sensor_statuses: typing.List[SensorHealth]
    """Statuses of the running lidars."""
    service_statuses: typing.List[NodeHealth]
    """Statuses of the services."""

    @staticmethod
    def _from_pb(pb: pm_pb.GetHealthStatusResponse) -> "HealthStatus":
        license_status: LicenseStatus = LicenseStatus._from_pb(pb.license_status)
        sensor_statuses: typing.List[SensorHealth] = []
        service_statuses: typing.List[NodeHealth] = []

        if LicensePrivilege.READ in license_status.privileges:
            sensor_statuses = [
                SensorHealth._from_pb(sensor_status)
                for sensor_status in pb.sensor_statuses
            ]
            service_statuses = [
                NodeHealth._from_pb(node_status) for node_status in pb.service_statuses
            ]

        return HealthStatus(
            license_status=license_status,
            sensor_statuses=sensor_statuses,
            service_statuses=service_statuses,
        )

File: cpp/horus/sdk/health.cpp
#include "horus/sdk/health.h"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <stdexcept>
#include <utility>
#include <vector>

#include "horus/logs/format.h"  // IWYU pragma: keep
#include "horus/pb/config/metadata_pb.h"
#include "horus/pb/cow.h"
#include "horus/pb/cow_repeated.h"
#include "horus/pb/license_server/messages_pb.h"
#include "horus/pb/preprocessing/messages_pb.h"
#include "horus/pb/project_manager/service_pb.h"
#include "horus/sdk/errors.h"
#include "horus/strings/str_cat.h"
#include "horus/strings/string_view.h"
#include "horus/types/span.h"

namespace horus {
namespace sdk {

LicenseStatus::LicenseInfo::LicenseInfo(const pb::LicenseInfo& pb_license_info)
    : expiration_timestamp_{std::chrono::system_clock::from_time_t(
                                pb_license_info.expiration_date().seconds()) +
                            std::chrono::nanoseconds{pb_license_info.expiration_date().nanos()}},
      number_of_lidars_{pb_license_info.lidar_count()} {
  allowed_features_.reserve(pb_license_info.allowed_features().size());
  for (const Cow<pb::LicenseInfo::AllowedFeature> allowed_feature :
       pb_license_info.allowed_features()) {
    allowed_features_.push_back(allowed_feature.Ref().feature());
  }
}

LicenseStatus::LicenseStatus(const pb::LicenseStatus& pb_license_status) : license_info_{} {
  switch (pb_license_status.license_level().level_case()) {
    case pb::LicenseLevel::LevelOneof::kLevel1: {
      license_level_ = 1;
      reason_ = StrCat(pb_license_status.license_level().level_1().license_server_unreachable());
      break;
    }
    case pb::LicenseLevel::LevelOneof::kLevel2: {
      license_level_ = 2;
      reason_ = StrCat(pb_license_status.license_level().level_2().license_not_found());
      break;
    }
    case pb::LicenseLevel::LevelOneof::kLevel3: {
      license_level_ = 3;
      switch (pb_license_status.license_level().level_3().reason_case()) {
        case pb::LicenseLevel_Level3::ReasonOneof::kLicenseExpired:
          reason_ = StrCat(pb_license_status.license_level().level_3().license_expired());
          break;
        case pb::LicenseLevel_Level3::ReasonOneof::kLicenseServerDisconnected:
          reason_ =
              StrCat(pb_license_status.license_level().level_3().license_server_disconnected());
          break;
        case pb::LicenseLevel_Level3::ReasonOneof::kNotSet:
        default:
          throw UnexpectedDataReceivedFromRpcError{
              StrCat("Unexpected level 3 reason sent by the RPC: ",
                     static_cast<std::uint32_t>(
                         pb_license_status.license_level().level_3().reason_case()))};
      }
      reason_ = StrCat(pb_license_status.license_level().level_2().license_not_found());
      break;
    }
    case pb::LicenseLevel::LevelOneof::kLevel4: {
      license_level_ = 4;
      reason_ = StrCat(pb_license_status.license_level().level_4().license_active());
      break;
    }
    case pb::LicenseLevel::LevelOneof::kNotSet:
    default: {
      throw sdk::UnexpectedDataReceivedFromRpcError{
          StrCat("Unexpected level sent by the RPC: ",
                 static_cast<std::uint32_t>(pb_license_status.license_level().level_case()))};
    }
  }

  if (pb_license_status.has_license_info()) {
    license_info_ = LicenseInfo{pb_license_status.license_info()};
  }

  const auto if_has_privilege_then_add_to_set =
      [this, privileges{pb_license_status.license_level().privilege()}](
          const Privilege privilege) -> void {
    if ((privileges & static_cast<std::uint32_t>(privilege)) ==
        static_cast<std::uint32_t>(privilege)) {
      privileges_.push_back(privilege);
    }
  };
  if_has_privilege_then_add_to_set(Privilege::kLowest);
  if_has_privilege_then_add_to_set(Privilege::kSubscribe);
  if_has_privilege_then_add_to_set(Privilege::kLicenseActions);
  if_has_privilege_then_add_to_set(Privilege::kRead);
  if_has_privilege_then_add_to_set(Privilege::kWrite);
}

const LicenseStatus::LicenseInfo& LicenseStatus::GetLicenseInfo() const {
  if (!HasLicenseInfo()) {
    throw std::logic_error("current license status does not carry a license info");
  }
  return license_info_;
}

SensorInfo::SensorInfo(const pb::SensorInfo& pb_sensor_info, bool is_unreachable)
    : lidar_id_{pb_sensor_info.lidar_id().Str()}, measured_frequency_{static_cast<double>(NAN)} {
  if (is_unreachable) {
    return;
  }

  std::vector<SensorStatus> statuses;
  const auto if_has_status_flag_then_add_to_set =
      [&statuses, status{pb_sensor_info.status()}](SensorStatus flag) -> void {
    if ((status & static_cast<std::uint32_t>(flag)) == static_cast<std::uint32_t>(flag)) {
      statuses.push_back(flag);
    }
  };
  if_has_status_flag_then_add_to_set(SensorStatus::kNoData);
  if_has_status_flag_then_add_to_set(SensorStatus::kReceivingData);
  if_has_status_flag_then_add_to_set(SensorStatus::kLowFrequency);
  if_has_status_flag_then_add_to_set(SensorStatus::kHighFrequency);
  if_has_status_flag_then_add_to_set(SensorStatus::kTilted);
  if_has_status_flag_then_add_to_set(SensorStatus::kObstructed);
  status_ = std::move(statuses);

  measured_frequency_ = pb_sensor_info.measured_frequency();
}

Span<const SensorInfo::SensorStatus> SensorInfo::Status() const noexcept {
  return {status_.begin().base(), status_.end().base()};
}

bool SensorInfo::HasStatus(const SensorStatus status) const {
  return std::find(status_.begin(), status_.end(), status) != status_.end();
}

StringView SensorHealth::NodeUnreachableErrorMessage() const {
  if (node_unreachable_error_message_.empty()) {
    throw std::logic_error("sensor status can be reached; no unreachable error message found");
  }
  return node_unreachable_error_message_;
}

HealthStatus::HealthStatus(const pb::GetHealthStatusResponse& pb_response)
    : license_status_{pb_response.license_status()} {
  // License level does not grant the Read privilege.
  if (!HasReadPrivilege()) {
    return;
  }

  // License grant the Read privilege: we have sensor and service statuses.
  sensor_statuses_.reserve(pb_response.sensor_statuses().size());
  for (const Cow<pb::GetHealthStatusResponse::SensorHealth>& sensor_health_status :
       pb_response.sensor_statuses()) {
    sensor_statuses_.emplace_back(sensor_health_status.Ref());
  }

  service_statuses_.reserve(pb_response.service_statuses().size());
  for (const Cow<pb::GetHealthStatusResponse::NodeHealth>& service_status :
       pb_response.service_statuses()) {
    service_statuses_.emplace_back(service_status.Ref());
  }
}

Span<const SensorHealth> HealthStatus::SensorStatuses() const {
  return {sensor_statuses_.begin().base(), sensor_statuses_.end().base()};
}

Span<const NodeHealth> HealthStatus::ServiceStatuses() const {
  return {service_statuses_.begin().base(), service_statuses_.end().base()};
}

}  // namespace sdk
}  // namespace horus